cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1240,11 +1250,141 @@ public long getProgramRunCount(ProgramId programId) {
     return result;
   }
 
+  void createStartTimeIfNotExist() {
+    MDSKey mdsKey = new MDSKey(RUN_COUNT_FIRST_UPGRADE_TIME);
+    if (get(mdsKey, Long.TYPE) == null) {
+      LOG.info("Start to upgrade the run count");
+      write(mdsKey, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()));
+    }
+  }
+
+  /**
+   * Compute the old run count if the run record is older than first start up time after 5.1
+   */
+  boolean computeOldRunCount(int maxRows) {
+    boolean upgradeDone = upgradeActiveRowkey(TYPE_RUN_RECORD_STARTING, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_STARTED, maxRows);
+    upgradeDone &= upgradeActiveRowkey(TYPE_RUN_RECORD_SUSPENDED, maxRows);
+    upgradeDone &= computeOldCompletedRunCount(maxRows);
+    return upgradeDone;
+  }
+
+  /**
+   * Upgrades the rowkeys for the active run record. Note that we do not count the run record of the old active runs
+   * since it is hard to deal with the races and we expect user to stop all the programs before they do the upgrade.
+   *
+   * @param recordType type of the record
+   * @param maxRows    maximum number of rows to be upgraded in this call.
+   * @return true if no rows required an upgrade
+   */
+  private boolean upgradeActiveRowkey(String recordType, int maxRows) {
+    LOG.info("Checking upgrade for {}", recordType);
+    MDSKey startKey = new MDSKey.Builder().add(recordType).build();
+    Map<MDSKey, RunRecordMeta> oldMap = listKV(startKey, RunRecordMeta.class, maxRows);
+
+    for (Map.Entry<MDSKey, RunRecordMeta> oldEntry : oldMap.entrySet()) {
+      MDSKey oldKey = oldEntry.getKey();
+      RunRecordMeta runRecord = oldEntry.getValue();
+      // extract the program id from the old key, the format will be same after the program name is reached, so the
+      // method can be used directly.
+      MDSKey newKey = getProgramKeyBuilder(TYPE_RUN_RECORD_ACTIVE, getProgramID(oldKey))
+        .add(runRecord.getStartTs())
+        .add(runRecord.getPid())
+        .build();
+
+      delete(oldKey);
+      write(newKey, runRecord);
+    }
+
+    LOG.info("Upgrading {} active run records of {}", oldMap.size(), recordType);
+
+    // If we are not able to get the max number of rows, that means we reach the end of this record type
+    return oldMap.size() < maxRows;
+  }
+
+  /**
+   * Upgrades the rowkeys for the completed run record which is older than the time at row key
+   * RUN_COUNT_FIRST_UPGRADE_TIME
+   *
+   * @param maxRows    maximum number of rows to be upgraded in this call.
+   * @return true if all old completed run records has been scanned
+   */
+  private boolean computeOldCompletedRunCount(int maxRows) {
+    LOG.info("Starting to compute the old run count");
```

Comment:
this message should be in the caller so that it's not repeated a bunch of times. 'Upgrading run counts'

Reply:
The upgrade() method in DefaultStore is called using a ScheduledService, so it will be hard to just log this once, as the upgrade() method itself will be called multiple times.

Reply:
I think that service should only be in charge of retrying if there are failures. It shouldn't be responsible for the number of runs to change per call. Also, have you tried with 1000 run records? Does it comfortably complete a batch of 1000 in one transaction?

Reply:
Yea, tried with over 1000, and it could do it. And 1000 is the value we used for previous app version update and the metadata upgrade